cpp/src/io/parquet/parquet.hpp

```diff
@@ -362,17 +362,111 @@ class CompactProtocolWriter {
   }
 
  public:
```

Comment:
can `put_xyz` members be private?
Edit: Actually, should they be moved to CompactProtocolWriterBuilder? CompactProtocolWriter can own the high level API, and CompactProtocolWriterBuilder can implement the field writing and below. As it is now, we have CPW calling into CPWB, which then calls CPW again for the low level code.

Reply:
In the same vein, is the fact that CompactProtocolWriter::write() is being called from CompactProtocolFieldWriter::field_struct_list() and CompactProtocolFieldWriter::field_struct() also an issue?
[Here](https://github.com/rapidsai/cudf/blob/96299361fc3cdbc9a99019289da71441cde2033b/cpp/src/io/parquet/parquet.hpp#L427) and [here](https://github.com/rapidsai/cudf/blob/96299361fc3cdbc9a99019289da71441cde2033b/cpp/src/io/parquet/parquet.hpp#L417).

Reply:
ugh, so it looks like there's really no layers of abstraction here. Maybe we're better off merging CompactProtocolWriterBuilder into CompactProtocolWriter.